File: sdks/typescript/src/graph.ts
import { Projection, QueryRequest, Traversal, readBatch } from "./dsl.js";

const PRIVATE_PREFIX = "__helix_graph_";
const NODE_ID = "__helix_graph_node_id";
const EXTERNAL_ID = "__helix_graph_external_id";
const NODE_LABEL = "__helix_graph_node_label";
const EDGE_ID = "__helix_graph_edge_id";
const EDGE_KEY = "__helix_graph_edge_key";
const EDGE_SOURCE = "__helix_graph_edge_source";
const EDGE_TARGET = "__helix_graph_edge_target";
const EDGE_LABEL = "__helix_graph_edge_label";
const EDGE_WEIGHT = "__helix_graph_edge_weight";

export type GraphDirection = "directed" | "undirected";
export type TraversalDirection = "out" | "in" | "both";
export type DegreeKind = "in" | "out" | "total";

export interface GraphSelectionOptions {
  nodeTraversal: Traversal<any, any>;
  edgeTraversal: Traversal<any, any>;
  direction?: GraphDirection;
  nodeProperties?: readonly string[];
  edgeProperties?: readonly string[];
  externalIdentityProperty?: string;
  graphifyEdgeKeyProperty?: string;
  weightProperty?: string;
  maxNodes?: number;
  maxEdges?: number;
  allowFullScan?: boolean;
}

/** Typed inputs for the one ordinary graph-loading read batch. */
export class GraphSelection {
  readonly nodeTraversal: Traversal<any, any>;
  readonly edgeTraversal: Traversal<any, any>;
  readonly direction: GraphDirection;
  readonly nodeProperties: readonly string[];
  readonly edgeProperties: readonly string[];
  readonly externalIdentityProperty?: string;
  readonly graphifyEdgeKeyProperty?: string;
  readonly weightProperty?: string;
  readonly maxNodes?: number;
  readonly maxEdges?: number;
  readonly allowFullScan: boolean;

  constructor(options: GraphSelectionOptions) {
    this.nodeTraversal = options.nodeTraversal;
    this.edgeTraversal = options.edgeTraversal;
    this.direction = options.direction ?? "directed";
    this.nodeProperties = [...new Set(options.nodeProperties ?? [])].sort();
    this.edgeProperties = [...new Set(options.edgeProperties ?? [])].sort();
    this.externalIdentityProperty = options.externalIdentityProperty;
    this.graphifyEdgeKeyProperty = options.graphifyEdgeKeyProperty;
    this.weightProperty = options.weightProperty;
    this.maxNodes = positiveInteger(options.maxNodes, "maxNodes");
    this.maxEdges = positiveInteger(options.maxEdges, "maxEdges");
    this.allowFullScan = options.allowFullScan ?? false;
    if (this.direction !== "directed" && this.direction !== "undirected") {
      throw new TypeError("direction must be 'directed' or 'undirected'");
    }
    for (const property of [
      ...this.nodeProperties,
      ...this.edgeProperties,
      this.externalIdentityProperty,
      this.graphifyEdgeKeyProperty,
      this.weightProperty,
    ]) {
      if (property === undefined) continue;
      if (property.length === 0) throw new TypeError("graph property names must not be empty");
      if (property.startsWith(PRIVATE_PREFIX)) throw new TypeError(`graph property uses reserved prefix: ${property}`);
    }
    const startsWithFullScan = (traversal: Traversal<any, any>): boolean => {
      const [source] = traversal.intoSteps();
      if (source === undefined || (source.variant !== "N" && source.variant !== "E")) return false;
      return (source.payload as { variant?: string } | undefined)?.variant === "All";
    };
    if (!this.allowFullScan && (startsWithFullScan(this.nodeTraversal) || startsWithFullScan(this.edgeTraversal))) {
      throw new TypeError("full graph scans require allowFullScan: true");
    }
  }

  toQueryRequest(): QueryRequest {
    const nodeProjection = [
      Projection.property("$id", NODE_ID),
      Projection.property(this.externalIdentityProperty ?? "$id", EXTERNAL_ID),
      Projection.property("$label", NODE_LABEL),
      ...this.nodeProperties.map((property) => Projection.property(property, property)),
    ];
    const edgeProjection = [
      Projection.property("$id", EDGE_ID),
      Projection.fromEndpoint("$id", EDGE_SOURCE),
      Projection.toEndpoint("$id", EDGE_TARGET),
      Projection.property("$label", EDGE_LABEL),
    ];
    if (this.graphifyEdgeKeyProperty !== undefined) edgeProjection.push(Projection.property(this.graphifyEdgeKeyProperty, EDGE_KEY));
    if (this.weightProperty !== undefined) edgeProjection.push(Projection.property(this.weightProperty, EDGE_WEIGHT));
    edgeProjection.push(...this.edgeProperties.map((property) => Projection.property(property, property)));
    const nodes = this.maxNodes === undefined ? this.nodeTraversal : this.nodeTraversal.limit(this.maxNodes + 1);
    const edges = this.maxEdges === undefined ? this.edgeTraversal : this.edgeTraversal.limit(this.maxEdges + 1);
    return readBatch()
      .varAs("nodes", nodes.project(nodeProjection))
      .varAs("edges", edges.project(edgeProjection))
      .returning(["nodes", "edges"])
      .toQueryRequest();
  }
}

export interface GraphNode {
  id: string;
  label?: string;
  attributes: Readonly<Record<string, unknown>>;
}

export interface GraphEdge {
  id: string;
  graphifyKey?: string;
  source: string;
  target: string;
  label?: string;
  weight?: number;
  attributes: Readonly<Record<string, unknown>>;
}

export interface BetweennessOptions {
  mode?: "exact" | "sampled" | "auto";
  sampleCount?: number;
  seed?: number;
  exactThrough?: number;
  normalized?: boolean;
  endpoints?: boolean;
  weighted?: boolean;
}

export interface TraversalOptions {
  seeds: readonly string[];
  maxDepth: number;
  strategy?: "breadthFirst" | "depthFirst";
  direction?: TraversalDirection;
  allowedLabels?: readonly string[];
  stopNonSeedAtOrAboveDegree?: number;
}

export interface LouvainOptions {
  resolution?: number;
  threshold?: number;
  seed?: number;
  maxLevels?: number;
}

export interface LayoutOptions {
  k?: number;
  iterations?: number;
  seed?: number;
  weighted?: boolean;
  initialPositions?: readonly { nodeId: string; x: number; y: number }[];
}

type NativeRecord = Record<string, any>;
type NativeGraphHandle = NativeRecord;

export interface NativeGraphModule extends NativeRecord {
  graph_from_query_response(spec: NativeRecord, response: Uint8Array): NativeGraphHandle;
}

export interface GraphLoadingClient {
  _graphResponse(request: QueryRequest, spec: NativeRecord): Promise<Uint8Array | NativeGraphHandle>;
}

/** Raised only when `client.graph(...)` is used without native bindings. */
export class NativeGraphUnavailable extends Error {
  constructor(message: string) {
    super(message);
    this.name = "NativeGraphUnavailable";
  }
}

/** Immutable graph whose accessors and algorithms all execute in Rust. */
export class NativeGraph {
  constructor(
    private readonly native: NativeGraphHandle,
    private readonly module: NativeGraphModule,
  ) {}

  get nodeCount(): number {
    return Number(this.native.node_count());
  }

  get edgeCount(): number {
    return Number(this.native.edge_count());
  }

  get directed(): boolean {
    return this.native.is_directed();
  }

  get multigraph(): boolean {
    return this.native.is_multigraph();
  }

  get attributes(): Readonly<Record<string, unknown>> {
    return decodeJson(this.native.graph_attributes_json());
  }

  containsNode(nodeId: string): boolean {
    return this.native.contains_node(nodeId);
  }

  containsEdge(edgeId: string): boolean {
    return this.native.contains_edge(edgeId);
  }

  nodes(): readonly GraphNode[] {
    return this.native.nodes().map(nodeRecord);
  }

  edges(): readonly GraphEdge[] {
    return this.native.edges().map(edgeRecord);
  }

  node(nodeId: string): GraphNode | undefined {
    const record = this.native.node(nodeId);
    return record === undefined ? undefined : nodeRecord(record);
  }

  edge(edgeId: string): GraphEdge | undefined {
    const record = this.native.edge(edgeId);
    return record === undefined ? undefined : edgeRecord(record);
  }

  neighbors(nodeId: string, direction: TraversalDirection = "both"): readonly string[] {
    return this.native.neighbors(nodeId, nativeDirection(this.module, direction));
  }

  successors(nodeId: string): readonly string[] {
    return this.native.successors(nodeId);
  }

  predecessors(nodeId: string): readonly string[] {
    return this.native.predecessors(nodeId);
  }

  outEdgeIds(nodeId: string): readonly string[] {
    return this.native.out_edge_ids(nodeId);
  }

  inEdgeIds(nodeId: string): readonly string[] {
    return this.native.in_edge_ids(nodeId);
  }

  incidentEdgeIds(nodeId: string): readonly string[] {
    return this.native.incident_edge_ids(nodeId);
  }

  edgesBetween(source: string, target: string, direction: TraversalDirection = "both"): readonly string[] {
    return this.native.edges_between(source, target, nativeDirection(this.module, direction));
  }

  hasEdgeBetween(source: string, target: string, direction: TraversalDirection = "both"): boolean {
    return this.native.has_edge_between(source, target, nativeDirection(this.module, direction));
  }

  degree(nodeId: string, kind: DegreeKind = "total"): NativeRecord {
    return this.native.degree(nodeId, nativeDegreeKind(this.module, kind));
  }

  degrees(kind: DegreeKind = "total"): readonly NativeRecord[] {
    return this.native.degrees(nativeDegreeKind(this.module, kind));
  }

  async betweennessCentrality(options: BetweennessOptions = {}): Promise<readonly NativeRecord[]> {
    return this.native.betweenness_centrality_async(nativeBetweennessOptions(this.module, options));
  }

  async edgeBetweennessCentrality(options: BetweennessOptions = {}): Promise<readonly NativeRecord[]> {
    return this.native.edge_betweenness_centrality_async(nativeBetweennessOptions(this.module, options));
  }

  async simpleCycles(lengthBound: number, maxCycles?: number): Promise<NativeRecord> {
    return this.native.simple_cycles_async(lengthBound, maxCycles);
  }

  async traverse(options: TraversalOptions): Promise<NativeRecord> {
    const strategy =
      (options.strategy ?? "breadthFirst") === "breadthFirst"
        ? this.module.NativeTraversalStrategy.BreadthFirst
        : this.module.NativeTraversalStrategy.DepthFirst;
    const hub =
      options.stopNonSeedAtOrAboveDegree === undefined
        ? this.module.NativeHubExpansionPolicy.ExpandAll()
        : this.module.NativeHubExpansionPolicy.StopNonSeedAtOrAbove(options.stopNonSeedAtOrAboveDegree);
    return this.native.traverse_async({
      strategy,
      seeds: [...options.seeds],
      max_depth: options.maxDepth,
      direction: nativeDirection(this.module, options.direction ?? "both"),
      allowed_labels: [...(options.allowedLabels ?? [])],
      hub_policy: hub,
    });
  }

  async shortestPath(
    source: string,
    target: string,
    options: { direction?: TraversalDirection; allowedLabels?: readonly string[]; maxDepth?: number } = {},
  ): Promise<NativeRecord> {
    return this.native.shortest_path_async(
      source,
      target,
      nativeDirection(this.module, options.direction ?? "both"),
      [...(options.allowedLabels ?? [])],
      options.maxDepth,
    );
  }

  async louvainCommunities(options: LouvainOptions = {}): Promise<NativeRecord> {
    return this.native.louvain_communities_async(
      options.resolution ?? 1,
      options.threshold ?? 1e-4,
      options.seed ?? 42,
      options.maxLevels ?? 10,
    );
  }

  async springLayout(options: LayoutOptions = {}): Promise<readonly NativeRecord[]> {
    return this.native.spring_layout_async(
      options.k,
      options.iterations ?? 50,
      options.seed ?? 42,
      options.weighted ?? true,
      (options.initialPositions ?? []).map((position) => ({ node_id: position.nodeId, x: position.x, y: position.y })),
    );
  }

  inducedSubgraph(nodeIds: readonly string[]): NativeGraph {
    return new NativeGraph(this.native.induced_subgraph([...nodeIds]), this.module);
  }

  toUndirected(): NativeGraph {
    return new NativeGraph(this.native.to_undirected(), this.module);
  }

  copy(): NativeGraph {
    return new NativeGraph(this.native.copy(), this.module);
  }

  compose(right: NativeGraph): NativeGraph {
    if (this.module !== right.module) throw new TypeError("graphs must originate from the same native binding module");
    return new NativeGraph(this.native.compose(right.native), this.module);
  }

  relabel(mapping: Readonly<Record<string, string>> | ReadonlyMap<string, string>): NativeGraph {
    const entries = mapping instanceof Map ? [...mapping] : Object.entries(mapping);
    return new NativeGraph(this.native.relabel(entries.map(([from, to]) => ({ from, to }))), this.module);
  }
}

export async function loadGraph(client: GraphLoadingClient, selection: GraphSelection): Promise<NativeGraph> {
  const native = await loadNativeGraphModule();
  const spec = {
    direction: selection.direction === "directed" ? native.NativeGraphDirection.Directed : native.NativeGraphDirection.Undirected,
    node_limit: selection.maxNodes,
    edge_limit: selection.maxEdges,
  };
  const response = await client._graphResponse(selection.toQueryRequest(), spec);
  return new NativeGraph(response instanceof Uint8Array ? native.graph_from_query_response(spec, response) : response, native);
}

const DEFAULT_NATIVE_PACKAGE = "@helix-db/uniffi";
const dynamicImport = new Function("specifier", "return import(specifier)") as (specifier: string) => Promise<NativeGraphModule>;

async function loadNativeGraphModule(): Promise<NativeGraphModule> {
  const packageName = process.env.HELIXDB_UNIFFI_NODE_PACKAGE ?? DEFAULT_NATIVE_PACKAGE;
  let native: NativeGraphModule;
  try {
    native = await dynamicImport(packageName);
  } catch (error) {
    throw new NativeGraphUnavailable(`native graph bindings unavailable: ${error instanceof Error ? error.message : String(error)}`);
  }
  for (const name of ["NativeGraphDirection", "NativeTraversalDirection", "graph_from_query_response"]) {
    if (native[name] === undefined) throw new NativeGraphUnavailable(`${packageName} does not export ${name}`);
  }
  return native;
}

function nodeRecord(record: NativeRecord): GraphNode {
  return { id: record.id, label: record.label, attributes: decodeJson(record.attributes_json) };
}

function edgeRecord(record: NativeRecord): GraphEdge {
  return {
    id: record.id,
    graphifyKey: record.graphify_key,
    source: record.source,
    target: record.target,
    label: record.label,
    weight: record.weight,
    attributes: decodeJson(record.attributes_json),
  };
}

function decodeJson(bytes: Uint8Array): Readonly<Record<string, unknown>> {
  return JSON.parse(new TextDecoder().decode(bytes)) as Record<string, unknown>;
}

function nativeDirection(native: NativeGraphModule, direction: TraversalDirection): unknown {
  const values = {
    out: native.NativeTraversalDirection.Out,
    in: native.NativeTraversalDirection.In,
    both: native.NativeTraversalDirection.Both,
  };
  return values[direction];
}

function nativeDegreeKind(native: NativeGraphModule, kind: DegreeKind): unknown {
  const values = { in: native.NativeDegreeKind.In, out: native.NativeDegreeKind.Out, total: native.NativeDegreeKind.Total };
  return values[kind];
}

function nativeBetweennessOptions(native: NativeGraphModule, options: BetweennessOptions): NativeRecord {
  const sampleCount = positiveInteger(options.sampleCount ?? 100, "sampleCount")!;
  const seed = options.seed ?? 42;
  const exactThrough = nonNegativeInteger(options.exactThrough ?? 1_000, "exactThrough");
  const mode = options.mode ?? "exact";
  return {
    mode:
      mode === "exact"
        ? native.NativeBetweennessMode.Exact()
        : mode === "sampled"
          ? native.NativeBetweennessMode.Sampled(sampleCount, seed)
          : native.NativeBetweennessMode.Auto(exactThrough, sampleCount, seed),
    normalized: options.normalized ?? true,
    endpoints: options.endpoints ?? false,
    weighted: options.weighted ?? false,
  };
}

function positiveInteger(value: number | undefined, name: string): number | undefined {
  if (value === undefined) return undefined;
  if (!Number.isSafeInteger(value) || value <= 0) throw new TypeError(`${name} must be a positive safe integer`);
  return value;
}

function nonNegativeInteger(value: number, name: string): number {
  if (!Number.isSafeInteger(value) || value < 0) throw new TypeError(`${name} must be a non-negative safe integer`);
  return value;
}

export { EDGE_SOURCE, EXTERNAL_ID };

File: sdks/go/graph.go
package helix

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"sort"
)

const (
	graphPrivatePrefix = "__helix_graph_"
	graphNodeID        = "__helix_graph_node_id"
	graphExternalID    = "__helix_graph_external_id"
	graphNodeLabel     = "__helix_graph_node_label"
	graphEdgeID        = "__helix_graph_edge_id"
	graphEdgeKey       = "__helix_graph_edge_key"
	graphEdgeSource    = "__helix_graph_edge_source"
	graphEdgeTarget    = "__helix_graph_edge_target"
	graphEdgeLabel     = "__helix_graph_edge_label"
	graphEdgeWeight    = "__helix_graph_edge_weight"
)

var ErrNativeGraphUnavailable = errors.New("helix: native graph bindings are not linked")

// GraphDirection controls whether algorithms preserve stored edge direction.
type GraphDirection uint8

const (
	GraphDirected GraphDirection = iota + 1
	GraphUndirected
)

// GraphSelection builds the single ordinary read used by Client.Graph.
type GraphSelection struct {
	NodeTraversal            *Traversal
	EdgeTraversal            *Traversal
	Direction                GraphDirection
	NodeProperties           []string
	EdgeProperties           []string
	ExternalIdentityProperty string
	GraphifyEdgeKeyProperty  string
	WeightProperty           string
	MaxNodes                 uint64
	MaxEdges                 uint64
	AllowFullScan            bool
}

func (s GraphSelection) request() (Request, graphLoadSpec, error) {
	if s.NodeTraversal == nil || s.EdgeTraversal == nil {
		return nil, graphLoadSpec{}, errors.New("helix: graph selection requires node and edge traversals")
	}
	if s.Direction != GraphDirected && s.Direction != GraphUndirected {
		return nil, graphLoadSpec{}, errors.New("helix: graph direction must be directed or undirected")
	}
	startsWithFullScan := func(traversal *Traversal) bool {
		steps := traversal.Steps()
		if len(steps) == 0 || (steps[0].kind != "N" && steps[0].kind != "E") {
			return false
		}
		switch reference := steps[0].value.(type) {
		case NodeRef:
			return reference.kind == "All"
		case EdgeRef:
			return reference.kind == "All"
		default:
			return false
		}
	}
	if !s.AllowFullScan && (startsWithFullScan(s.NodeTraversal) || startsWithFullScan(s.EdgeTraversal)) {
		return nil, graphLoadSpec{}, errors.New("helix: full graph scans require AllowFullScan")
	}
	properties := append(append([]string{}, s.NodeProperties...), s.EdgeProperties...)
	properties = append(properties, s.ExternalIdentityProperty, s.GraphifyEdgeKeyProperty, s.WeightProperty)
	for _, property := range properties {
		if property == "" {
			continue
		}
		if len(property) >= len(graphPrivatePrefix) && property[:len(graphPrivatePrefix)] == graphPrivatePrefix {
			return nil, graphLoadSpec{}, fmt.Errorf("helix: graph property uses reserved prefix: %s", property)
		}
	}
	nodeProperties := uniqueSorted(s.NodeProperties)
	edgeProperties := uniqueSorted(s.EdgeProperties)
	nodeProjection := []Projection{
		ProjectPropAs("$id", graphNodeID),
		ProjectPropAs(valueOr(s.ExternalIdentityProperty, "$id"), graphExternalID),
		ProjectPropAs("$label", graphNodeLabel),
	}
	for _, property := range nodeProperties {
		if property == "" {
			return nil, graphLoadSpec{}, errors.New("helix: graph property names must not be empty")
		}
		nodeProjection = append(nodeProjection, ProjectPropAs(property, property))
	}
	edgeProjection := []Projection{
		ProjectPropAs("$id", graphEdgeID),
		ProjectFromEndpoint("$id", graphEdgeSource),
		ProjectToEndpoint("$id", graphEdgeTarget),
		ProjectPropAs("$label", graphEdgeLabel),
	}
	if s.GraphifyEdgeKeyProperty != "" {
		edgeProjection = append(edgeProjection, ProjectPropAs(s.GraphifyEdgeKeyProperty, graphEdgeKey))
	}
	if s.WeightProperty != "" {
		edgeProjection = append(edgeProjection, ProjectPropAs(s.WeightProperty, graphEdgeWeight))
	}
	for _, property := range edgeProperties {
		if property == "" {
			return nil, graphLoadSpec{}, errors.New("helix: graph property names must not be empty")
		}
		edgeProjection = append(edgeProjection, ProjectPropAs(property, property))
	}
	nodes := TraversalFromSteps(s.NodeTraversal.Steps())
	edges := TraversalFromSteps(s.EdgeTraversal.Steps())
	if s.MaxNodes > 0 {
		nodes.Limit(s.MaxNodes + 1)
	}
	if s.MaxEdges > 0 {
		edges.Limit(s.MaxEdges + 1)
	}
	request := NewReadQueryRequest(
		Read().
			VarAs("nodes", nodes.Project(nodeProjection...)).
			VarAs("edges", edges.Project(edgeProjection...)).
			Returning("nodes", "edges"),
	)
	return request, graphLoadSpec{
		Direction:        s.Direction,
		ExternalIdentity: s.ExternalIdentityProperty != "",
		GraphifyEdgeKey:  s.GraphifyEdgeKeyProperty != "",
		NodeLimit:        optionalLimit(s.MaxNodes),
		EdgeLimit:        optionalLimit(s.MaxEdges),
	}, nil
}

type graphLoadSpec struct {
	Direction        GraphDirection
	ExternalIdentity bool
	GraphifyEdgeKey  bool
	NodeLimit        *uint64
	EdgeLimit        *uint64
}

// Graph executes exactly one ordinary read and returns a reusable native graph.
func (c *Client) Graph(ctx context.Context, selection GraphSelection) (*NativeGraph, error) {
	request, spec, err := selection.request()
	if err != nil {
		return nil, err
	}
	if !nativeGraphAvailable() {
		return nil, ErrNativeGraphUnavailable
	}
	requestBytes, err := MarshalRequest(request)
	if err != nil {
		return nil, &HelixError{Kind: ErrorSerialization, Err: err, Details: err.Error()}
	}
	if c != nil && c.embedded != nil {
		if direct, ok := c.embedded.(interface {
			Graph([]byte, graphLoadSpec) (graphBackend, error)
		}); ok {
			backend, err := direct.Graph(requestBytes, spec)
			if err != nil {
				return nil, &HelixError{Kind: ErrorEmbedded, Err: err, Details: err.Error()}
			}
			return &NativeGraph{backend: backend}, nil
		}
	}
	var response json.RawMessage
	if err := c.Exec(ctx, request, &response); err != nil {
		return nil, err
	}
	backend, err := graphFromQueryResponse(spec, response)
	if err != nil {
		return nil, err
	}
	return &NativeGraph{backend: backend}, nil
}

// GraphNode keeps selected attributes as lazy JSON bytes.
type GraphNode struct {
	ID             string
	Label          *string
	AttributesJSON []byte
}

func (n GraphNode) Attributes() (map[string]any, error) {
	var attributes map[string]any
	err := json.Unmarshal(n.AttributesJSON, &attributes)
	return attributes, err
}

// GraphEdge preserves stable Helix identity and optional Graphify key.
type GraphEdge struct {
	ID             string
	GraphifyKey    *string
	Source         string
	Target         string
	Label          *string
	Weight         *float64
	AttributesJSON []byte
}

func (e GraphEdge) Attributes() (map[string]any, error) {
	var attributes map[string]any
	err := json.Unmarshal(e.AttributesJSON, &attributes)
	return attributes, err
}

type BetweennessMode uint8

const (
	BetweennessExact BetweennessMode = iota + 1
	BetweennessSampled
	BetweennessAuto
)

type BetweennessOptions struct {
	Mode         BetweennessMode
	SampleCount  uint64
	Seed         uint64
	ExactThrough uint64
	Normalized   bool
	Endpoints    bool
	Weighted     bool
}

func GraphifyBetweennessOptions() BetweennessOptions {
	return BetweennessOptions{Mode: BetweennessAuto, SampleCount: 100, Seed: 42, ExactThrough: 1_000, Normalized: true}
}

type NodeScore struct {
	NodeID string
	Score  float64
}
type EdgeScore struct {
	EdgeID      string
	GraphifyKey *string
	Source      string
	Target      string
	Score       float64
}
type Cycle struct {
	NodeIDs []string
	EdgeIDs []string
}
type CycleResult struct {
	Cycles    []Cycle
	Truncated bool
}

type TraversalDirection uint8

const (
	TraversalOut TraversalDirection = iota + 1
	TraversalIn
	TraversalBoth
)

type TraversalStrategy uint8

const (
	TraversalBreadthFirst TraversalStrategy = iota + 1
	TraversalDepthFirst
)

type TraversalOptions struct {
	Strategy                   TraversalStrategy
	Seeds                      []string
	MaxDepth                   uint64
	Direction                  TraversalDirection
	AllowedLabels              []string
	StopNonSeedAtOrAboveDegree *uint64
}

type Visit struct {
	NodeID         string
	Depth          uint64
	DiscoveryOrder uint64
}
type EdgeTraversalDirection uint8

const (
	EdgeTraversalForward EdgeTraversalDirection = iota + 1
	EdgeTraversalReverse
)

type TraversedEdge struct {
	EdgeID             string
	GraphifyKey        *string
	Source             string
	Target             string
	TraversalDirection EdgeTraversalDirection
	Label              *string
}
type TraversalResult struct {
	Visits         []Visit
	DiscoveryEdges []TraversedEdge
}

type DegreeKind uint8

const (
	DegreeIn DegreeKind = iota + 1
	DegreeOut
	DegreeTotal
)

type NodeDegree struct {
	NodeID         string
	Degree         uint64
	WeightedDegree float64
}
type PathEdge struct {
	EdgeID             string
	GraphifyKey        *string
	Source             string
	Target             string
	TraversalDirection EdgeTraversalDirection
	Label              *string
	AttributesJSON     []byte
}
type PathResultKind uint8

const (
	PathMissingSource PathResultKind = iota + 1
	PathMissingTarget
	PathNoPath
	PathFound
)

type PathResult struct {
	Kind    PathResultKind
	NodeIDs []string
	Edges   []PathEdge
}
type LouvainOptions struct {
	Resolution float64
	Threshold  float64
	Seed       uint64
	MaxLevels  uint64
}
type Community struct {
	ID      string
	NodeIDs []string
}
type CommunityResult struct {
	Communities []Community
	Modularity  float64
	Levels      uint64
}
type NodePosition struct {
	NodeID string
	X      float64
	Y      float64
}
type LayoutOptions struct {
	K                *float64
	Iterations       uint64
	Seed             uint64
	Weighted         bool
	InitialPositions []NodePosition
}

// NativeGraph is immutable; algorithms never query Helix after construction.
type NativeGraph struct{ backend graphBackend }

func (g *NativeGraph) NodeCount() uint64                  { return g.backend.NodeCount() }
func (g *NativeGraph) EdgeCount() uint64                  { return g.backend.EdgeCount() }
func (g *NativeGraph) IsDirected() bool                   { return g.backend.IsDirected() }
func (g *NativeGraph) IsMultigraph() bool                 { return g.backend.IsMultigraph() }
func (g *NativeGraph) AttributesJSON() ([]byte, error)    { return g.backend.AttributesJSON() }
func (g *NativeGraph) ContainsNode(id string) bool        { return g.backend.ContainsNode(id) }
func (g *NativeGraph) ContainsEdge(id string) bool        { return g.backend.ContainsEdge(id) }
func (g *NativeGraph) Nodes() ([]GraphNode, error)        { return g.backend.Nodes() }
func (g *NativeGraph) Edges() ([]GraphEdge, error)        { return g.backend.Edges() }
func (g *NativeGraph) Node(id string) (*GraphNode, error) { return g.backend.Node(id) }
func (g *NativeGraph) Edge(id string) (*GraphEdge, error) { return g.backend.Edge(id) }
func (g *NativeGraph) Neighbors(id string, direction TraversalDirection) ([]string, error) {
	return g.backend.Neighbors(id, direction)
}
func (g *NativeGraph) Successors(id string) ([]string, error)   { return g.backend.Successors(id) }
func (g *NativeGraph) Predecessors(id string) ([]string, error) { return g.backend.Predecessors(id) }
func (g *NativeGraph) OutEdgeIDs(id string) ([]string, error)   { return g.backend.OutEdgeIDs(id) }
func (g *NativeGraph) InEdgeIDs(id string) ([]string, error)    { return g.backend.InEdgeIDs(id) }
func (g *NativeGraph) IncidentEdgeIDs(id string) ([]string, error) {
	return g.backend.IncidentEdgeIDs(id)
}
func (g *NativeGraph) EdgesBetween(source, target string, direction TraversalDirection) ([]string, error) {
	return g.backend.EdgesBetween(source, target, direction)
}
func (g *NativeGraph) HasEdgeBetween(source, target string, direction TraversalDirection) (bool, error) {
	return g.backend.HasEdgeBetween(source, target, direction)
}
func (g *NativeGraph) Degree(id string, kind DegreeKind) (NodeDegree, error) {
	return g.backend.Degree(id, kind)
}
func (g *NativeGraph) Degrees(kind DegreeKind) []NodeDegree { return g.backend.Degrees(kind) }
func (g *NativeGraph) BetweennessCentrality(options BetweennessOptions) ([]NodeScore, error) {
	return g.backend.BetweennessCentrality(options)
}
func (g *NativeGraph) EdgeBetweennessCentrality(options BetweennessOptions) ([]EdgeScore, error) {
	return g.backend.EdgeBetweennessCentrality(options)
}
func (g *NativeGraph) SimpleCycles(lengthBound uint64, maxCycles *uint64) (CycleResult, error) {
	return g.backend.SimpleCycles(lengthBound, maxCycles)
}
func (g *NativeGraph) Traverse(options TraversalOptions) (TraversalResult, error) {
	return g.backend.Traverse(options)
}
func (g *NativeGraph) ShortestPath(source, target string, direction TraversalDirection, labels []string, maxDepth *uint64) (PathResult, error) {
	return g.backend.ShortestPath(source, target, direction, labels, maxDepth)
}
func (g *NativeGraph) LouvainCommunities(options LouvainOptions) (CommunityResult, error) {
	return g.backend.LouvainCommunities(options)
}
func (g *NativeGraph) SpringLayout(options LayoutOptions) ([]NodePosition, error) {
	return g.backend.SpringLayout(options)
}
func (g *NativeGraph) InducedSubgraph(ids []string) (*NativeGraph, error) {
	backend, err := g.backend.InducedSubgraph(ids)
	if err != nil {
		return nil, err
	}
	return &NativeGraph{backend: backend}, nil
}
func (g *NativeGraph) ToUndirected() (*NativeGraph, error) {
	backend, err := g.backend.ToUndirected()
	if err != nil {
		return nil, err
	}
	return &NativeGraph{backend: backend}, nil
}
func (g *NativeGraph) Copy() *NativeGraph { return &NativeGraph{backend: g.backend.Copy()} }
func (g *NativeGraph) Compose(right *NativeGraph) (*NativeGraph, error) {
	backend, err := g.backend.Compose(right.backend)
	if err != nil {
		return nil, err
	}
	return &NativeGraph{backend: backend}, nil
}
func (g *NativeGraph) Relabel(mapping map[string]string) (*NativeGraph, error) {
	backend, err := g.backend.Relabel(mapping)
	if err != nil {
		return nil, err
	}
	return &NativeGraph{backend: backend}, nil
}

type graphBackend interface {
	NodeCount() uint64
	EdgeCount() uint64
	IsDirected() bool
	IsMultigraph() bool
	AttributesJSON() ([]byte, error)
	ContainsNode(string) bool
	ContainsEdge(string) bool
	Nodes() ([]GraphNode, error)
	Edges() ([]GraphEdge, error)
	Node(string) (*GraphNode, error)
	Edge(string) (*GraphEdge, error)
	Neighbors(string, TraversalDirection) ([]string, error)
	Successors(string) ([]string, error)
	Predecessors(string) ([]string, error)
	OutEdgeIDs(string) ([]string, error)
	InEdgeIDs(string) ([]string, error)
	IncidentEdgeIDs(string) ([]string, error)
	EdgesBetween(string, string, TraversalDirection) ([]string, error)
	HasEdgeBetween(string, string, TraversalDirection) (bool, error)
	Degree(string, DegreeKind) (NodeDegree, error)
	Degrees(DegreeKind) []NodeDegree
	BetweennessCentrality(BetweennessOptions) ([]NodeScore, error)
	EdgeBetweennessCentrality(BetweennessOptions) ([]EdgeScore, error)
	SimpleCycles(uint64, *uint64) (CycleResult, error)
	Traverse(TraversalOptions) (TraversalResult, error)
	ShortestPath(string, string, TraversalDirection, []string, *uint64) (PathResult, error)
	LouvainCommunities(LouvainOptions) (CommunityResult, error)
	SpringLayout(LayoutOptions) ([]NodePosition, error)
	InducedSubgraph([]string) (graphBackend, error)
	ToUndirected() (graphBackend, error)
	Copy() graphBackend
	Compose(graphBackend) (graphBackend, error)
	Relabel(map[string]string) (graphBackend, error)
}

func uniqueSorted(values []string) []string {
	result := append([]string{}, values...)
	sort.Strings(result)
	return compact(result)
}
func compact(values []string) []string {
	if len(values) == 0 {
		return values
	}
	result := values[:1]
	for _, value := range values[1:] {
		if value != result[len(result)-1] {
			result = append(result, value)
		}
	}
	return result
}
func valueOr(value, fallback string) string {
	if value == "" {
		return fallback
	}
	return value
}
func optionalLimit(value uint64) *uint64 {
	if value == 0 {
		return nil
	}
	return &value
}
